File: CSharp/SpecStudioExamplesTestsCSharp/common/CartInputString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class CartInputString
    {
        public string totalItems;
        public string shipping;
        public string discount;
        public string totalPrice;
        public string notes;

        public CartInputString(string totalItems, string shipping, string discount, string totalPrice, string notes)
        {
            this.totalItems = totalItems;
            this.shipping = shipping;
            this.discount = discount;
            this.totalPrice = totalPrice;
            this.notes = notes;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static CartInputString FromText(string text)
        {
            var parts = Tokens.Require(text, 5, "CartInput");
            return new CartInputString(parts[0], parts[1], parts[2], parts[3], parts[4]);
        }

        public CartInputString(string text)
        {
            var parsed = FromText(text);
            this.totalItems = parsed.totalItems;
            this.shipping = parsed.shipping;
            this.discount = parsed.discount;
            this.totalPrice = parsed.totalPrice;
            this.notes = parsed.notes;
        }

        public CartInputTyped ToCartInputTyped()
        {
            return new CartInputTyped(
                new Dollar(this.totalItems),
                new Dollar(this.shipping),
                new Dollar(this.discount),
                new Dollar(this.totalPrice),
                this.notes
            );
        }

        public override string ToString()
        {
            return Tokens.Token(totalItems) + " " + Tokens.Token(shipping) + " " + Tokens.Token(discount) + " " + Tokens.Token(totalPrice) + " " + Tokens.Token(notes);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not CartInputString other) return false;
            return (DNCString == this.totalItems || DNCString == other.totalItems || object.Equals(this.totalItems, other.totalItems))
                && (DNCString == this.shipping || DNCString == other.shipping || object.Equals(this.shipping, other.shipping))
                && (DNCString == this.discount || DNCString == other.discount || object.Equals(this.discount, other.discount))
                && (DNCString == this.totalPrice || DNCString == other.totalPrice || object.Equals(this.totalPrice, other.totalPrice))
                && (DNCString == this.notes || DNCString == other.notes || object.Equals(this.notes, other.notes));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.totalItems);
            h.Add(this.shipping);
            h.Add(this.discount);
            h.Add(this.totalPrice);
            h.Add(this.notes);
            return h.ToHashCode();
        }
    }
}

File: JavaScript/ExampleTests/common/CartInputString.js
import * as tokens from "./tokens.js";

export class CartInputString {
  static DNC_STRING = "?DNC?";

  constructor(totalItems = "", shipping = "", discount = "", totalPrice = "", notes = "") {
    this.totalItems = totalItems;
    this.shipping = shipping;
    this.discount = discount;
    this.totalPrice = totalPrice;
    this.notes = notes;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new CartInputString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? "",
      v[3] ?? "",
      v[4] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 5, "CartInput");
    return new CartInputString(
      parts[0],
      parts[1],
      parts[2],
      parts[3],
      parts[4]
    );
  }

  toString() {
    return tokens.token(this.totalItems) + " " + tokens.token(this.shipping) + " " + tokens.token(this.discount) + " " + tokens.token(this.totalPrice) + " " + tokens.token(this.notes);
  }

  equals(other) {
    if (!(other instanceof CartInputString)) return false;
    return (this.totalItems === CartInputString.DNC_STRING || other.totalItems === CartInputString.DNC_STRING || this.totalItems === other.totalItems)
      && (this.shipping === CartInputString.DNC_STRING || other.shipping === CartInputString.DNC_STRING || this.shipping === other.shipping)
      && (this.discount === CartInputString.DNC_STRING || other.discount === CartInputString.DNC_STRING || this.discount === other.discount)
      && (this.totalPrice === CartInputString.DNC_STRING || other.totalPrice === CartInputString.DNC_STRING || this.totalPrice === other.totalPrice)
      && (this.notes === CartInputString.DNC_STRING || other.notes === CartInputString.DNC_STRING || this.notes === other.notes);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}
